File: Data Structures & Algorithms/search-in-rotated-sorted-array-ii/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l <= r:
            m = l + (r - l)//2
            
            # best case target is mid
            if target == nums[m]:
                return True
            # if can't decide sorted -> shrink range
            elif nums[l] == nums[m] == nums[r]:
                l += 1
                r -= 1
            # check what is sorted
            # left sorted
            elif nums[l] <= nums[m]:
                # means target in left -> move left
                if nums[l] <= target < nums[m]:
                    r = m - 1
                # target not in left -> move right
                else:
                    l = m + 1
            # right sorted
            # right > mid
            else:
                # target > mid -> target in right of mid -> move right
                if nums[r] >= target > nums[m]:
                    l = m + 1
                # target not in right -> move left
                else:
                    r = m - 1
        return False

File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-2.rs
impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut minimum_subarray_length = nums.len() + 1;
        let mut left = 0;
        let mut right = 0;
        let mut current_sum = 0;
        while right < nums.len(){
            current_sum += nums[right];

            while current_sum >= target {
                minimum_subarray_length = minimum_subarray_length.min(right - left + 1);
                current_sum -= nums[left];
                left += 1;
            }
            right += 1;
        }
        if minimum_subarray_length < nums.len() + 1 { minimum_subarray_length as i32 } else { 0i32 }
    }
}
